lib/AST/Type.cpp

```diff
@@ -2968,36 +2968,31 @@ Type TypeBase::getSuperclassForDecl(const ClassDecl *baseClass,
   llvm_unreachable("no inheritance relationship between given classes");
 }
 
-bool TypeBase::canTreatContextAsMember(const DeclContext *dc) {
-  Type baseTy(getRValueType());
-
-  if (auto metaBase = baseTy->getAs<AnyMetatypeType>()) {
-    baseTy = metaBase->getInstanceType()->getRValueType();
-  }
+static bool
+isContextContainedNominalTypeFromSuperClass(Type &baseTy, const DeclContext *dc,
+                                            LazyResolver *resolver) {
 
-  if (!baseTy->getAnyNominal())
+  // If baseTy is not a class type, return false;
+  if (!baseTy->getClassOrBoundGenericClass())
     return false;
 
-  // If the context is a protocol or its extensions, the base type should conform
-  // to that protocol.
-  if (auto Prot = dc->getAsProtocolOrProtocolExtensionContext()) {
-    for (auto Conf : baseTy->getAnyNominal()->getAllConformances()) {
-      if (Conf->getProtocol() == Prot)
-        return true;
+  // Collect all nominal type contexts that contain the given dc.
+  llvm::SmallPtrSet<const NominalTypeDecl*, 4> Parent;
```

Comment:
Ah, if you need to do this to fix it, the real problem is that you're calling getMemberSubstitutions() with the wrong dc.

The intent of that method is that it's called as part of qualified lookup; if we're coming from an unqualified lookup, we will have already done the walk up the parents in resolveType().

So if your member is a nested nominal type, you must be passing in the dc of the *nested* type, when you should be passing in the dc that *contains* the nested type.

Does this make sense? I still feel like the original assertion was correct...